Updating Number.Formatting to properly print -0

tannergooding opened the pull request:
This resolves https://github.com/dotnet/coreclr/issues/10290

tannergooding commented:
FYI. @danmosemsft, @jkotas 

This covers part of `IEEE 754:2008 - 5.12.1: External character sequences representing zeros, infinities, and NaNs`:
> The conversions (described in 5.4.2) from supported formats to external character sequences and back that recover the original floating-point representation, shall recover zeros, infinities, and quiet NaNs, as well as non-zero finite numbers. In particular, signs of zeros and infinities are preserved

tannergooding commented:
This fixes `System.Double` and `System.Single`. It also fixes `System.Decimal`, as that was the simpler change and I would believe is the desired behavior here as well.

This impacts 'R', as well as the other formats. If we want to restrict this to just 'R', that will be a slightly more involved change.

tannergooding commented:
Looks like `linux-musl` is using a C runtime without `signbit` support, will fix.

-- Looking at CoreFX there doesn't appear to be existing code coverage for "R" or some of the special-cases (tests are still running locally to confirm I didn't just miss them). I'll work on getting those cases covered.

tannergooding commented on src/pal/src/cruntime/math.cpp:
```diff
@@ -453,6 +453,34 @@ PALIMPORT double __cdecl PAL_pow(double x, double y)
+x  Double-precision floating-point value
+
+--*/
+int __cdecl PAL_signbit(double x)
```
Had to add `PAL_signbit`.... Clang complains about importing `fp.h` and using `FPDOUBLE` to get the sign-bit, because the header also declares a static function which is unused in `grisu3.cpp`

Rather than fix the header or compilation flags, I opted to just use `signbit`, since it is the standard library function to do this.

tannergooding replied:
Adding it to `palinternal.h` doesn't work in this case, because it is a macro in the `math.h` file.

I had to do what the other "macro" PAL redefinitions are doing (like `isfinite` and `isnan`), which is to define a new function (`_signbit`) which calls the macro, and then use the new function everywhere.
For the existing ones, the new function name was selected to match the MSVC function name (`_finite`, `_isnan`, etc). Since there is no `_signbit` function for MSVC (just `signbit`), I had to (for MSVC) also do `#define signbit _signbit`.

tannergooding commented:
Will go through the various CoreFX failures and make sure they are all expected.
Will then prep the PR which fixes them to be "correct" for after this is merged and gets pumped over.

tannergooding commented:
At least one of the failures is due to a bad change (specifically relating to integers). The break is in `RoundNumber`, which is used for rounding `integers`, `decimals`, and `doubles`/`floats`.

We are currently bypassing rounding (for `decimal`) in certain cases by passing an additional `isDecimal` parameter. There are a few ways I could fix this to keep the current behavior for `integer` and the new behavior for `double`/`float`. However, given that `double`/`float` also have special rounding rules (which `RoundNumber` doesn't currently handle), and that is may be good to generally track the original number "kind", I am working on a change that would allow `NumberBuffer` to carry around the "kind" (`Unknown`, `Integer`, `Decimal`, or `Double`) so that we can appropriately handle the various behaviors where necessary.

tannergooding commented:
`System.Memory.Tests` failures are because the UTF8String formatter also needs to be updated.

tannergooding commented:
Rebased to try and get jobs to run, most of them didn't kick off last time.

CoreFX side change is here: https://github.com/dotnet/corefx/pull/32109

eerhardt commented on src/System.Private.CoreLib/shared/System/Number.Formatting.cs:
```diff
@@ -488,6 +490,7 @@ private static string FormatSingle(ref ValueStringBuilder sb, float value, ReadO
             char fmt = ParseFormatSpecifier(format, out int digits);
             int precision = FloatPrecision;
             NumberBuffer number = default;
+            number.kind = NumberBufferKind.Double;
```
Would it make more sense for `NumberBufferKind.Double` to be renamed to `NumberBufferKind.FloatingPoint` or similar?  It feels odd to use `Double` when we are working with `float`/`Single`

tannergooding replied:
Possibly, but I would like to do that as a follow-up PR.

Currently we use `DoubleToNumber` for both `single` and `double`.... We should probably either rename that as well, or add an explicit code-path for `SingleToNumber`

eerhardt replied:
We are introducing the `enum` in this PR.  Why not give it a good name from the start?

tannergooding replied:
Because, until the the methods are split to handle `Single` explicitly, it is actually a `NumberBuffer` which contains a `Double` (`Single` is upcast in order to make the call) and so the name is more accurate as-is.

eerhardt replied:
Ah, got it. That makes sense.

tannergooding commented:
> This comment seems out of date now. I don't see you updating the number.h file

Fixed, added comments clarifying the expected offset of each field as well.

tannergooding commented:
test Ubuntu x64 Checked Innerloop Build and Test please
test Ubuntu x64 Checked Innerloop Build and Test (Jit - TieredCompilation=0) please
test OSX10.12 x64 Checked Innerloop Build and Test please

tannergooding commented:
Turns out the Windows x86 implementation was using a custom implementation written in assembly (that still used the x87 FPU stack).
I've removed that in the latest commit, in favor of the shared implementation under `number.cpp` (which was already used everywhere else).

tannergooding commented:
test Ubuntu x64 Checked jitstress1
test Ubuntu x64 Checked jitstress2

test Windows_NT x64 Checked jitstress1
test Windows_NT x64 Checked jitstress2

test Windows_NT x86 Checked jitstress1
test Windows_NT x86 Checked jitstress2

tannergooding commented:
@danmosemsft, @jkotas, @eerhardt.

I believe all feedback has been resolved, but the PR has changed a bit since some of your approvals. I'd appreciate it if you could give it another cursory glance and sign-off again if everything looks good.

danmosemsft commented:
Seems reasonable but this is not my area. It would be good for @jkotas to OK it.

jkotas commented:
The invalidated CoreFX tests need to be disabled to make the CI green.

tannergooding commented:
@jkotas, is there a file tracking those tests in CoreCLR, or do they need to be disabled in CoreFX directly?

jkotas commented:
Add them to: https://github.com/dotnet/coreclr/blob/master/tests/CoreFX/CoreFX.issues.json

tannergooding commented:
test Ubuntu x64 Checked jitstress1
test Ubuntu x64 Checked jitstress2

test Windows_NT x64 Checked jitstress1
test Windows_NT x64 Checked jitstress2

test Windows_NT x86 Checked jitstress1
test Windows_NT x86 Checked jitstress2

tannergooding commented:
test Ubuntu x64 Checked jitstress1
test Ubuntu x64 Checked jitstress2

test Windows_NT x64 Checked jitstress1
test Windows_NT x64 Checked jitstress2

test Windows_NT x86 Checked jitstress1
test Windows_NT x86 Checked jitstress2

tannergooding commented:
Fixed the name of the CoreFX tests in the json file.

tannergooding commented:
@fiigii, the `Avx2.Gather` failures are unrelated. Could you take a look?

tannergooding commented:
@jkotas, any other feedback?

If not, this should be good to merge, the only failures are in the JitStress1 and JitStress2 jobs and are unrelated to the PR.

fiigii commented:
@tannergooding I will look into this failure tomorrow, thanks.
